Research/ResearchUI/iOS/Views/RSDVerticalGridCollectionView.swift

```diff
@@ -0,0 +1,191 @@
+        }
+    }
+    
+    /// The ratio of cellSpacing to cell horizontal border
+    open var cellHorizontalBorderRatio = 0.5
+    
+    /// The number of items in the grid
+    open var itemCount: Int = 0 {
+        didSet {
+            if itemCount > 0 {
+                self.itemSize = self.cellSize(for: IndexPath(row: 0, section: 0))
+            }
+        }
+    }
+    
+    /// This needs to be used for the collection view delegate section count
+    public var sectionCount: Int {
+        return ((itemCount - 1) / columnCount) + 1
+    }
+    
+    /// This needs to be used for the collection view delegate section count
+    public func itemCountInSection(section: Int) -> Int {
+        let lastRowIndex = itemCount / columnCount
+        if section < lastRowIndex {
+            return columnCount
+        }
+        return itemCount % columnCount
+    }
+    
+    /// Return the item index from the index path
+    public func itemIndex(for indexPath: IndexPath) -> Int {
+        return (indexPath.section * columnCount) + indexPath.row
+    }
+    
+    /// The leading and trailing padding for a cell to make it centered horizontally
+    /// For most items in most rows, this is 0, as they are a full row,
+    /// but for the last row where there may be less items than the column count,
+    /// this can be used to make those items the same size as the previous items rows
+    public func seciontInset(for section: Int) -> UIEdgeInsets {
+        
+        // Check for valid collection view size
+        let collectionViewWidth = (collectionView?.bounds.width ?? 0)
+        if collectionViewWidth == 0 {
+            return UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 0)
+        }
+
+        // See if we are in the last row, the row of interest
+        let lastRowIndex = itemCount / columnCount
+
+        // Top inset is zero on first row, and cell spacing for every other
+        var topInset = verticalCellSpacing
+        if section == 0 {
+            topInset = 0
+        }
+        // Bottom inset is zero for all row but the last, and may be set again below
+        var bottomInset = CGFloat(0)
+        
+        // Default insets of the cell spacing
+        // Check for item being in the last row
+        if section >= lastRowIndex {
+            bottomInset = verticalCellSpacing
+        }
+        // Calculate the cells width in the row and find the section inset to center them
+        let itemsInRow = itemCountInSection(section: section)
+        let cellsWidth = (self.cellWidth * CGFloat(itemsInRow)) + (horizontalCellSpacing * CGFloat(itemsInRow - 1))
+        let leadingTrailingInset = CGFloat(Int((collectionViewWidth - cellsWidth) * CGFloat(0.5))) - 1
+        
+        return UIEdgeInsets(top: topInset, left: leadingTrailingInset, bottom: bottomInset, right: leadingTrailingInset)
+    }
+    
+    /// The cell size for the item
+    public func cellSize(for indexPath: IndexPath) -> CGSize {
```

Comment:
why is this taking the index path as an input if you aren't making it open? Also, if there is a reason for it *not* being open, then should be `public final`.

Reply:
Made it public final